File: Week 2 Arrays/Problem Sets/Problem Set 2/readability/readability.c
/**
 * @file readability.c
 * @brief CS50x Problem Set 2 — Readability: Coleman-Liau grade estimator.
 *
 * @author Ayman Elmasry — AEL Digital Studio
 * @project AEL Sovereign — CS50x 2026-2027
 *
 * @details
 *   Estimates the U.S. school grade level of a body of text using the
 *   Coleman-Liau readability index. The program counts letters, words and
 *   sentences, then maps those counts through the classic formula:
 *
 *       index = 0.0588 * L - 0.296 * S - 15.8
 *
 *   where L = letters per 100 words and S = sentences per 100 words.
 *   The rounded index is reported as a grade, with the two contract
 *   clamps: indices below 1 print "Before Grade 1" and indices at or
 *   above 16 print "Grade 16+".
 *
 * Algorithm:
 *   1. Read the text with get_string.
 *   2. Count letters (alphabetic), words (whitespace-delimited tokens)
 *      and sentences (terminated by '.', '!' or '?').
 *   3. Derive the per-100-word rates and evaluate the formula.
 *   4. Clamp and print the grade.
 *
 * Complexity:
 *   Time  — O(n): each character is examined a constant number of times
 *           across the three counting passes.
 *   Space — O(1): only scalar counters, independent of text length.
 */

#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

/* Coleman-Liau formula coefficients and contract clamps. */
#define LETTERS_COEF 0.0588
#define SENTENCES_COEF 0.296
#define FORMULA_OFFSET 15.8
#define PER_WORDS_SCALE 100.0
#define GRADE_FLOOR 1
#define GRADE_CEILING 16

/**
 * @brief Count the alphabetic letters in a string.
 *
 * @param text The text to scan.
 * @return The number of characters for which isalpha is true.
 */
static int count_letters(string text)
{
    int letters = 0;

    for (int i = 0; text[i] != '\0'; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }

    return letters;
}

/**
 * @brief Count the words in a string as whitespace-delimited tokens.
 *
 * @param text The text to scan.
 * @return The number of words detected.
 */
static int count_words(string text)
{
    int words = 0;
    bool inside_word = false;

    for (int i = 0; text[i] != '\0'; i++)
    {
        if (isspace(text[i]))
        {
            inside_word = false;
        }
        else if (!inside_word)
        {
            /* The first non-space character after a break begins a word. */
            inside_word = true;
            words++;
        }
    }

    return words;
}

/**
 * @brief Count the sentences in a string.
 *
 * @param text The text to scan.
 * @return The number of '.', '!' or '?' characters found.
 */
static int count_sentences(string text)
{
    int sentences = 0;

    for (int i = 0; text[i] != '\0'; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
    }

    return sentences;
}

/**
 * @brief Compute the (unclamped) Coleman-Liau index for a text.
 *
 * @param letters   Number of alphabetic characters.
 * @param words     Number of words.
 * @param sentences Number of sentences.
 * @return The rounded readability index.
 */
static int coleman_liau_grade(int letters, int words, int sentences)
{
    /* Express both rates relative to a fixed block of 100 words. */
    double letters_per_100 = (double) letters / words * PER_WORDS_SCALE;
    double sentences_per_100 = (double) sentences / words * PER_WORDS_SCALE;

    double index = LETTERS_COEF * letters_per_100
                 - SENTENCES_COEF * sentences_per_100
                 - FORMULA_OFFSET;

    return (int) round(index);
}

int main(void)
{
    string text = get_string("Text: ");

    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    int index = coleman_liau_grade(letters, words, sentences);

    if (index < GRADE_FLOOR)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= GRADE_CEILING)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }

    return 0;
}

File: Week 6 Python/Problem Sets/Problem Set 6/sentimental-readability/readability.py
"""
AEL Sovereign — CS50x 2026-2027
Problem Set 6: Sentimental Readability
Author: Ayman Elmasry — AEL Digital Studio

Estimates the U.S. school grade level required to understand a passage of
text, using the Coleman-Liau index.  The user is prompted for a sentence
or paragraph and the computed grade level is printed.
"""


def count_letters(text):
    """Return the number of alphabetic characters in the text."""
    return sum(1 for character in text if character.isalpha())


def count_words(text):
    """Return the number of words in the text, split on whitespace."""
    return len(text.split())


def count_sentences(text):
    """Return the number of sentences, based on '.', '!', and '?'."""
    return sum(1 for character in text if character in ".!?")


def grade_level(letters, words, sentences):
    """
    Compute the Coleman-Liau reading grade.  L is the average number of
    letters per 100 words and S is the average number of sentences per
    100 words; the index is 0.0588 * L - 0.296 * S - 15.8.
    """
    per_100 = 100 / words
    letters_per = letters * per_100
    sentences_per = sentences * per_100
    return 0.0588 * letters_per - 0.296 * sentences_per - 15.8


def report(index):
    """
    Render the reading level as a string.  Indexes below 1 are reported
    as "Before Grade 1", indexes of 16 or higher as "Grade 16+", and any
    other value is rounded to the nearest whole grade.
    """
    if index < 1:
        return "Before Grade 1"
    if index >= 16:
        return "Grade 16+"
    return f"Grade {round(index)}"


def main():
    """Prompt for text and print the estimated reading grade level."""
    text = input("Text: ")
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)
    index = grade_level(letters, words, sentences)
    print(report(index))


if __name__ == "__main__":
    main()
